update the import docs (#6104)
ollama/ollama: docs/import.md

# Importing a model

> You can import a model into Ollama;

Reviewer: ```suggestion
You can import a model into Ollama:
```

  * From Safetensors weights; or

> 

Reviewer: this newline makes the bullets very spaced out

  * From a GGUF file

> ## Importing from Safetensors weights

Reviewer: ```suggestion
## Importing from Safetensors
```

Importing your model directly from the Safetensors weights is the easiest way to
import a model. Ollama supports importing several different architectures including:

  * Llama (including Llama 2, Llama 3, and Llama 3.1);
  * Mistral (including Mistral 1, Mistral 2, and Mixtral); and
  * Gemma (including Gemma 1 and Gemma 2)

> This includes importing either the foundation model as well as any fine tuned model which

Reviewer: we may want to remove these arbitrary newlines for better readability

is based upon one of these architectures.

To import a model, create a Modelfile using a text editor which includes a `FROM`
line which points to the directory with the model weights.

> ```dockerfile

Reviewer: do we want to break this out into:

1. Step 1: Write a Modelfile
2. Step 2: Run `ollama create`

Author: I really like this suggestion.

# Importing a model

You can import a model into Ollama:

  * From Safetensors weights; or
  * From a GGUF file

## Importing from Safetensors

> Importing your model directly from the Safetensors weights is the easiest way to import a model. Ollama supports importing several different architectures including:

Reviewer: It may be clearer to start with the instructions here:

First, create a `Modelfile`

```
FROM /path/to/safetensors
```

Then run `ollama create`:

```
ollama create my-model
```

Lastly, test the model

```
ollama run my-model
```

and then have a supported model heading below:

#### Supported models

* Llama (including Llama 2, Llama 3, and Llama 3.1);
* Phi (I think we support this now too)
* Mistral (including Mistral 1, Mistral 2, and Mixtral); and
* Gemma (including Gemma 1 and Gemma 2)

>   * Llama (including Llama 2, Llama 3, and Llama 3.1);

Reviewer: Do we want to provide the architecture names? e.g. `LlamaForCausalLM`

Author: No, I think they just add confusion. I'm not sure any normal user would know what that meant.

Reviewer: 👍

>   * Llama (including Llama 2, Llama 3, and Llama 3.1);

Reviewer: nit: maybe a matter of style but I rarely see semicolons (vs nothing) at the end of list items if there are bullets

Author: I'm following the Plain English style guide: https://www.plainenglish.co.uk/how-to-write-in-plain-english.html 

> With a list that is part of a continuous sentence, put semicolons (;) after each point and start each with a lower-case letter.

  * Mistral (including Mistral 1, Mistral 2, and Mixtral); and
  * Gemma (including Gemma 1 and Gemma 2)

This includes importing either the foundation model as well as any fine tuned model which is based upon one of these architectures.

Importing a model requires three steps:

  1. Write a Modelfile
  2. Create your model with `ollama create my-model`
  3. Test your model with `ollama run my-model`

#### Writing a Modelfile

Create a Modelfile using a text editor which includes a `FROM` line which points to the directory with the model weights.

```dockerfile
FROM /path/to/safetensors/directory
```

If you create the Modelfile in the same directory as the weights, you can use the line `FROM .`.

You can include other settings in your Modelfile such as parameters for the model, the desired chat template, the system prompt, and any license information. Refer to the [Modelfile documentation](https://github.com/ollama/ollama/blob/main/docs/modelfile.md) for more information about Modelfile commands.

#### Create your model with `ollama create`

After you have created the Modelfile, use the command:

```bash
$ ollama create -f Modelfile <model name>
```

> If the model architecture of your weights is supported, Ollama will take a few moments to import the weights. It will also attempt to find the correct _chat template_ inside the model's configuration data. You can override the template using the `TEMPLATE` command in your Modelfile.

Reviewer: ```suggestion
If the model architecture of your weights is supported, Ollama will take a few moments to import the weights. It will also attempt to find the correct _chat template_ inside the model's `tokenizer_config.json` file. You can override the template using the `TEMPLATE` command in your Modelfile.
```

Author: I purposely removed this from the original document because I don't think it's useful. We don't tell you how to modify the `tokenizer_config.json` file to get it to pick up the template correctly, nor would we want to (since you can just using the `TEMPLATE` command to do that). Ultimately the mechanism isn't important here.

Reviewer: I agree - makes sense!

#### Test your model with `ollama run`

Once you have created a model, you can run it using the command:

```bash
$ ollama run <model name>
```

This will load the newly created model into memory and you can test to make certain that it's working correctly. You may have to make changes to your Modelfile and run the `ollama create` command again in order to get everything correct.

> After it is working, if you have created an account on [ollama.com](https://ollama.com), you can push your model to [ollama.com](https://ollama.com) to share it with other people. Use the name `<user name>/<model name>` (e.g. `jmorganca/my-model`) when you are creating the model, or use the `ollama cp` command to name the model with your user name at the beginning of the model name. You can then push the model using the command:

Reviewer: This is great – should we put it under a `Push your model to ollama.com` header?

```bash
$ ollama push <user name>/<model name>
```

> ## Importing a GGUF based model

Reviewer: ```suggestion
## Importing from GGUF
```

If you have a GGUF based model it is possible to import it into Ollama. You can obtain a GGUF model by:

>   * Converting a Safetensors model with the `convert_hf_to_gguf.py` from Llama.cpp; or

Reviewer: Could we omit given the safetensors support above?

  * Downloading a model from a place such as HuggingFace

To import the GGUF file, create a Modelfile:

```dockerfile

> FROM /path/to/file.gguf

Reviewer: This is missing the simple copy+paste instructions that safetensors has above. Maybe unnecessary but thought I'd point it out

# Importing a model

> You can import a model or fine tuned adapter into Ollama:

Reviewer: Let's add a ToC at the top here with three links the headers so people can jump to the right one:

* Import from GGUF
* Import from Safetensors
* Import from Safetensors Adapter
* Sharing models on ollama.com

  * Hugging Face [fine tuning framework] (https://huggingface.co/docs/transformers/en/training)
  * [Unsloth](https://github.com/unslothai/unsloth)
  * [MLX](https://github.com/ml-explore/mlx)

## Importing a model from Safetensors weights

First, create a `Modelfile` with a `FROM` command which points to the directory containing your Safetensors weights:

```dockerfile
FROM /path/to/safetensors/directory
```

If you create the Modelfile in the same directory as the weights, you can use the command `FROM .`.

Now run the `ollama create` command from the directory where you created the `Modelfile`:

```shell
ollama create my-model
```

Lastly, test the model:

```shell
ollama run my-model
```

Ollama supports importing models for several different architectures including:

  * Llama (including Llama 2, Llama 3, and Llama 3.1);
  * Mistral (including Mistral 1, Mistral 2, and Mixtral);
  * Gemma (including Gemma 1 and Gemma 2); and
  * Phi3

This includes importing foundation models as well as any fine tuned models which which have been _fused_ with a foundation model.

> ## Importing a GGUF based model

Reviewer: We lost the `ollama create` command for this one so it's harder to follow now

If you have a GGUF based model or adapter it is possible to import it into Ollama. You can obtain a GGUF model or adapter by:

  * converting a Safetensors model with the `convert_hf_to_gguf.py` from Llama.cpp; 
  * converting a Safetensors adapter with the `convert_lora_to_gguf.py` from Llama.cpp; or
  * downloading a model or adapter from a place such as HuggingFace

To import a GGUF model, create a `Modelfile` containg:

```dockerfile
FROM /path/to/file.gguf
```

For a GGUF adapter, create the `Modelfile` with:

```dockerfile
FROM <model name>
ADAPTER /path/to/file.gguf
```

When importing a GGUF adapter, it's important to use the same base model as the base model that the adapter was created with. You can use:

 * a model from Ollama
 * a GGUF file
 * a Safetensors based model 

## Quantizing a Model

> Quantizing a model allows you to run models faster and with less memory consumption but at reduced accuracy. This allows you to run a model on more modest hardware.

Reviewer: Needs an example

```
ollama create my-model --quantize q4_0
```

Reviewer: Ah, I see one below the lists of quantizations, but I think 80% of folks may be looking for the command.

Ollama can quantize FP16 and FP32 based models into different quantization levels using the `-q/--quantize` flag with the `ollama create` command.

### Supported Quantizations

> - `Q4_0`

Reviewer: Maybe out of scope of this change but should these be lowercase the way tags are on ollama.com?

- `Q5_K_M`
- `Q6_K`

First, create a Modelfile with the FP16 or FP32 based model you wish to quantize.

```dockerfile
FROM /path/to/my/gemma/f16/model
```

Use `ollama create` to then create the quantized model.

```shell
$ ollama create -q Q4_K_M mymodel
transferring model data
quantizing F16 model to Q4_K_M
creating new layer sha256:735e246cc1abfd06e9cdcf95504d6789a6cd1ad7577108a70d9902fef503c1bd
creating new layer sha256:0853f0ad24e5865173bbf9ffcc7b0f5d56b66fd690ab1009867e45e7d2c4db0f
writing manifest
success
```

> ## Sharing your model on [ollama.com](https://ollama.com)

Reviewer: Nice addition
